server/chat-plugins/teams.ts

```diff
@@ -0,0 +1,686 @@
+/**
+ * Plugin for sharing / storing teams in a database.
+ * By Mia.
+ * @author mia-pi-git
+ */
+
+import {PostgresDatabase, FS, Utils} from '../../lib';
+
+/** Maximum amount of teams a user can have stored at once. */
+const MAX_TEAMS = 200;
+/** Max teams that can be viewed in a search */
+const MAX_SEARCH = 3000;
+
+export interface StoredTeam {
+	teamid: number;
+	team: string;
+	ownerid: ID;
+	format: ID;
+	title: string | null;
+	date: Date;
+	private: boolean;
+	views: number;
+}
+
+interface TeamSearch {
+	format?: string;
+	owner?: string;
+	pokemon?: string[];
+	moves?: string[];
+	abilities?: string[];
+	gen?: number;
+}
+
+function refresh(context: Chat.PageContext) {
+	return (
+		`<button class="button" name="send" value="/j ${context.pageid}" style="float: right">` +
+		` <i class="fa fa-refresh"></i> ${context.tr('Refresh')}</button>`
+	);
+}
+
+export const TeamsHandler = new class {
+	database = new PostgresDatabase();
+	readyPromise: Promise<void> | null = Config.usepostgres ? (async () => {
+		try {
+			await this.database.query('SELECT * FROM teams LIMIT 1');
+		} catch {
+			await this.database.query(FS(`databases/schemas/teams.sql`).readSync());
+		}
+	})() : null;
+	destroy() {
+		void this.database.destroy();
+	}
+
+	async search(search: TeamSearch, user: User, count = 10, includePrivate = false) {
+		const args = [];
+		const where = [];
+		if (search.format) {
+			where.push(`format = $${args.length + 1}`);
+			args.push(toID(search.format));
+		}
+		if (search.owner) {
+			where.push(`ownerid = $${args.length + 1}`);
+			args.push(toID(search.owner));
+		}
+		if (search.gen) {
+			where.push(`format LIKE 'gen${search.gen}%'`);
+		}
+		if (!includePrivate) where.push('private != true');
+
+		const result = await this.query<StoredTeam>(
+			`SELECT * FROM teams${where.length ? ` WHERE ${where.join(' AND ')}` : ''} LIMIT ${count}`,
+			args,
+		);
+		return result.filter(row => {
+			const team = Teams.unpack(row.team)!;
+			if (row.private && !(row.ownerid === user.id || user.can('rangeban'))) {
```

Comment:
Why should admins be able to view a user's private teams?

Reply:
Admins can view private replays, I just applied the same principle here.

Reply:
There's a few reasons why admins should be able to view private replays like if a user is boosting on the ladder but I don't see a reason why admins should be able to view private teams since they are pretty sensitive information.

Reply:
Fair enough. Done.